### PR TITLE
feat(Modelarts): add a new resource to manage model

### PR DESCRIPTION
**Pull request by hmbig2**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
Running tool: /usr/local/go/bin/go test -timeout 9000000000s -run ^TestAccModelartsModel_basic$ github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/modelarts -v

=== RUN   TestAccModelartsModel_basic
=== PAUSE TestAccModelartsModel_basic
=== CONT  TestAccModelartsModel_basic
--- PASS: TestAccModelartsModel_basic (44.04s)
PASS
ok  	github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/modelarts	44.079s

```

### CODE REVIEW

File: docs/resources/modelarts_model.md [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+* `name` - (Required, String, ForceNew) Model name, which consists of 1 to 64 characters.  
+  Only letters, digits, hyphens (-), and underscores (_) are allowed.
+
+  Changing this parameter will create a new resource.
+
+* `model_type` - (Optional, String, ForceNew) Model type.  
+  It can be **TensorFlow**, **MXNet**, **Caffe**, **Spark_MLlib**, **Scikit_Learn**,
+  **XGBoost**, **Image**, **PyTorch**, or **Template** read from the configuration file.
+
+  Changing this parameter will create a new resource.
```
Comment (Jason-Zhang9309) [inline docs/resources/modelarts_model.md]: please move this part to line 311

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+}
+
+func modelartsModelModelDocsSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"doc_url": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `HTTP(S) link of the document.`,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:233]: add forcenew here, in list, if the parameter of element changed, that will not cause the whole list change

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+}
+
+func modelartsModelTemplateSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"infer_format": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `ID of the input and output mode.`,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:255]: diito

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+	return &sc
+}
+
+func modelartsModelTemplateTemplateInputSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"input": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Template input path, which can be a path to an OBS file or directory.`,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:282]: diito

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+	return &sc
+}
+
+func modelartsModelDependencySchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"installer": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Installation mode. Only **pip** is supported.`,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:302]: diito

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+}
+
+func modelartsModelDependencyPackageSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"package_version": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `Version of a dependency package.`,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:324]: diito

File: docs/resources/modelarts_model.md [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+* `dependencies` - (Optional, List, ForceNew) Package required for inference code and model.  
+  If the package is read from the configuration file, this parameter can be left blank.
+
+  Changing this parameter will create a new resource.
+  The [Dependency](#ModelartsModel_Dependency) structure is documented below.
+
+<a name="ModelartsModel_ModelDocs"></a>
+The `ModelDocs` block supports:
+
+* `doc_url` - (Optional, String) HTTP(S) link of the document.
```
Comment (Jason-Zhang9309) [inline docs/resources/modelarts_model.md]: add forcenew and "changing this..."

File: docs/resources/modelarts_model.md [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+
+* `doc_name` - (Optional, String) Document name, which must start with a letter. Enter 1 to 48 characters.  
+  Only letters, digits, hyphens (-), and underscores (_) are allowed.
+
+<a name="ModelartsModel_Template"></a>
+The `Template` block supports:
+
+* `template_id` - (Required, String) ID of the used template.
```
Comment (Jason-Zhang9309) [inline docs/resources/modelarts_model.md]: diito

File: docs/resources/modelarts_model.md [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+* `template_inputs` - (Required, List) Template input configuration, specifying the source path for configuring a model.
+  The [TemplateInput](#ModelartsModel_TemplateInput) structure is documented below.
+
+* `infer_format` - (Optional, String) ID of the input and output mode.  
+  When this parameter is used, the input and output mode built in the template does not take effect.
+
+<a name="ModelartsModel_TemplateInput"></a>
+The `TemplateInput` block supports:
+
+* `input` - (Required, String) Template input path, which can be a path to an OBS file or directory.
```
Comment (Jason-Zhang9309) [inline docs/resources/modelarts_model.md]: diito

File: docs/resources/modelarts_model.md [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+  When you use a template with multiple input items to create a model,
+  if the target paths input_properties specified in the template are the same,
+  the OBS directory or OBS file name entered here must be unique to prevent files from being overwritten.
+
+* `input_id` - (Required, String) Input item ID, which is obtained from template details.
+
+<a name="ModelartsModel_Dependency"></a>
+The `Dependency` block supports:
+
+* `installer` - (Required, String) Installation mode. Only **pip** is supported.
```
Comment (Jason-Zhang9309) [inline docs/resources/modelarts_model.md]: diito

File: docs/resources/modelarts_model.md [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+
+* `packages` - (Required, List) Collection of dependency packages.
+  The [package](#ModelartsModel_package) structure is documented below.
+
+<a name="ModelartsModel_package"></a>
+The `package` block supports:
+
+* `package_version` - (Optional, String) Version of a dependency package.
```
Comment (Jason-Zhang9309) [inline docs/resources/modelarts_model.md]: diito

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 12]
```diff
@@ -0,0 +1,726 @@
...
+		ReadContext:   resourceModelartsModelRead,
+		DeleteContext: resourceModelartsModelDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(20 * time.Minute),
+		},
+		Schema: map[string]*schema.Schema{
+			"region": {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:39]: please adjust the order of the arguments, like the doc, to make this code easier to read

Comment (hmbig2) [reply huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:39]: ok

File: docs/resources/modelarts_model.md (hunk 1 above)
Comment (hmbig2) [reply docs/resources/modelarts_model.md]: OK，keep same order with schema

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go (hunk 2 above)
Comment (hmbig2) [reply huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:233]: added

File: docs/resources/modelarts_model.md (hunk 7 above)
Comment (hmbig2) [reply docs/resources/modelarts_model.md]: added

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 13]
```diff
@@ -0,0 +1,738 @@
...
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `Model name, which consists of 1 to 64 characters.`,
+			},
+			"model_type": {
+				Type:        schema.TypeString,
+				Optional:    true,
```
Comment (ShiChangkuo) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go]: according to the API, this parameter should be required. why we design it to be optional?

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 14]
```diff
@@ -0,0 +1,738 @@
...
+			"infer_format": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				Description: `ID of the input and output mode.`,
+			},
+			"template_id": {
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (ShiChangkuo) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:259]: can we develop a data source to query the template ID?

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go [hunk 15]
```diff
@@ -0,0 +1,738 @@
...
+}
+
+func createModelWaitingForStateCompleted(ctx context.Context, d *schema.ResourceData, meta interface{}, t time.Duration) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"PENDING"},
+		Target:  []string{"COMPLETED"},
+		Refresh: func() (interface{}, string, error) {
+			cfg := meta.(*config.Config)
+			region := cfg.GetRegion(d)
+			// createModelWaiting: missing operation notes
```
Comment (ShiChangkuo) [inline huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go]: this line should be deleted

File: docs/resources/modelarts_model.md [hunk 16]
```diff
@@ -0,0 +1,510 @@
...
+
+* `publishable_flag` - Whether a model is subscribed from AI Gallery.  
+
+* `created_at` - Time when a model is created, in milliseconds calculated from 1970.1.1 0:0:0 UTC.  
+
+## Timeouts
+
+This resource provides the following timeouts configuration options:
+
+* `create` - Default is 20 minute.
```
Comment (ShiChangkuo) [inline docs/resources/modelarts_model.md]:
```suggestion
* `create` - Default is 20 minutes.
```

File: docs/resources/modelarts_model.md [hunk 17]
```diff
@@ -0,0 +1,510 @@
...
+
+  Changing this parameter will create a new resource.
+
+* `model_docs` - (Optional, List, ForceNew) List of model description documents. A maximum of three documents are supported.
+
+  Changing this parameter will create a new resource.
+  The [ModelDocs](#ModelartsModel_ModelDocs) structure is documented below.
+
+* `template` - (Optional, List, ForceNew) Configuration items in a template.  
+  This parameter is mandatory when **model_type** is set to **Template**.
```
Comment (ShiChangkuo) [inline docs/resources/modelarts_model.md]:
```suggestion
  This parameter is mandatory when `model_type` is set to **Template**.
```

For arguments, we should use "`", please update them in others.

File: docs/resources/modelarts_model.md [hunk 18]
```diff
@@ -0,0 +1,510 @@
...
+
+* `id` - The resource ID.
+
+* `schema_doc` - Download address of the model schema file.
+
+* `image_address` - Image path generated after model packaging.
+
+* `model_size` - Model size, in bytes.
+
+* `model_status` - Model status.
```
Comment (ShiChangkuo) [inline docs/resources/modelarts_model.md]: suggest to use `status` which is more better.

Comment (hmbig2) [reply docs/resources/modelarts_model.md]: done

File: docs/resources/modelarts_model.md (hunk 16 above)
Comment (hmbig2) [reply docs/resources/modelarts_model.md]: done

File: docs/resources/modelarts_model.md (hunk 17 above)
Comment (hmbig2) [reply docs/resources/modelarts_model.md]: done

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go (hunk 15 above)
Comment (hmbig2) [reply huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go]: done

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go (hunk 13 above)
Comment (hmbig2) [reply huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go]: fixed

File: huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go (hunk 14 above)
Comment (hmbig2) [reply huaweicloud/services/modelarts/resource_huaweicloud_modelarts_model.go:259]: yes

**Pull request merged**